boltdesignsystem/bolt
# Add spread support to BoltElement + demo usage in Link and Button components

## Jira
- http://vjira2:8080/browse/BDS-2068

## Summary
Updates BoltElement to support using the React-like `...props` spread operator (courtesy of [open-wc](https://open-wc.org/developing/lit-helpers.html)!) + demos the huge gains via applying this to the Button and Link web components.

Once the updates from https://github.com/boltdesignsystem/bolt/pull/1795 land, even more boilerplate templating logic should be able to be removed!

## Details

1. Beyond just exporting the `spread` function in our `@bolt/element` package, this PR takes advantage of this new `spread` helper by applying it to eliminate all of duplicate web component templating logic that was previously required to retain any initial HTML attributes on specific `rootElement` tags (`<a>`, `<button>`, etc).

![image](https://user-images.githubusercontent.com/1617209/77755904-96c69c00-7004-11ea-9285-577a8b0a10a5.png)

![image](https://user-images.githubusercontent.com/1617209/77756405-76e3a800-7005-11ea-8e3c-4c30c8844de9.png)


2. The updates from the previous bullet also address the core cause of the issue (see linked Jira ticket) regarding the Twig-rendered `<bolt-button>` being unable to switch between rendering a `<button>` tag internally vs `<a>` tag if the web component is rendering using an initial root HTML element. 
    - Side note: because there isn't a straightforward way to switch a DOM node's tagName, the old approach can't readily be "fixed" -- whereas this new approach to handling rootElement rendering rendering approach doesn't run into this limitation (ie.  fixes the issue by using a different approach)

3. This PR also adds 4 additional sets of Jest tests to Button to be extra sure that rendering / re-rendering / internal behavior is all works exactly as expected (without requiring really any new logic to be written for Button or Link)

## How to test
- Review code changes
- Confirm the root but from [BDS-2068](http://vjira2:8080/browse/BDS-2068?filter=-1) is fixed with these updates
- Confirm if there's enough demo / docs material with this PR to get the ball rolling?
